### DISCUSSION TITLE
Sending messages is slow in large groups

### DISCUSSION THREAD

**Discussion by xAbdaziz** [tulir/whatsmeow#309] [WhatsApp protocol Q&A] [answered]
When I send a message in a group with 100+ members it takes 5-10 seconds to send, meanwhile in small groups & PM it's sent instantly.
I'm using the example code in godocs.

I also noticed that there's warnings spam in my console when I send a message
`[Client WARN] Failed to fetch prekey for 123456789.0:15@s.whatsapp.net: invalid prekey in prekey response: prekey node doesn't contain ID tag`

Can someone please tell me what should I do to fix this issue?

**Comment by mamur-rezeki** [comment]
Also looking for the solution.

**Comment by mamur-rezeki** [comment]
So I've migrate to postgresql database (SSD). Simple text message could take ± 4 seconds with 266 group participants. It is similar with SQLite database on SSD storage, take ± 6 seconds for same message and group. While SQLite on HDD is pretty slow for ± 55 seconds.

Three scenarios have tested on initial message send and secondary message if in this case they have prekey caching (or not).

I don't know if then Message node encryption is using databases to store the prekey or Identity and get it for encrypting 1 message node for every number participants.  I'm running on 150Mbps network speed.

**Comment by rull05** [comment]
any solution?

**Comment by ghost** [comment]
Encrypting a message for a large group is very slow. So we need to avoid it being encrypted by whatsmeow .
The current solution is limiting the participants in this list 

https://github.com/tulir/whatsmeow/blob/350073db195c6b6fae832a5b420b5b67fd39f6be/send.go#L539-L542

just like this :
```go
participants = []types.JID{ownID}
```

**Reply by ghost** [reply]
Yes, that a thing.

I already solved it by replacing `libsignal` library with my own fork.

https://github.com/tulir/libsignal-protocol-go/issues/2

**Reply by PBatista** [reply]
So just changing this part of not receiving participants all the time tends to improve performance? Can't you implement a cache or something like that?

**Reply by ghost** [reply]
So, as far as I know, when a message is sent to a group chat, by default WhatsMeow will fetch all group participants and then retrieve all participant devices known to our WhatsMeow client based on the participant list.

For example, if there are 100 participants and each participant has 3 devices that have interacted in the group, then:

100 * 3 = 300 devices (300 keys)

Thus, one chat message we send will be created and encrypted with each different key (300 keys).

The encryption process takes a lot of time due to several factors, such as bottlenecks in the speed of getting/storing keys (R/W) with the database used.

Actually, all the process estimates can be seen on `whatsmeow.SendResponse`.

> What I did here is replace the participant list with a list that only contains the client's own JID (`ownJID`) (`send.go#sendGroup()`).

> Therefore, the message will only be encrypted for our own devices.

Next, I thought that participants might request again to our client to fetch the message we sent earlier. Then the client will resend the message with the requester’s key. 

Now, during the re-sending process, because the request process may be very fast and numerous, calculated based on active participants in the group chat, for example, 60 participants. 

If the R/W database is overwhelmed by the goroutine process and causes "Database Locked," then some keys might not be stored and also not available when the client tries to resend the message with unavailable keys. 

This is where the "Segmentation error nil" appears because `libsignal` does not check if the key is available or not.

Some users might encounter this message description:

"Waiting for this message"

**Reply by PBatista** [reply]
I made the changes and actually, the sending process was really quick... I'm only worried about the keys and shipping issues... in this case, is it possible to continue doing this sending process and then in the backgroud or something like that? , did he try to get these keys that he tried to send? And another, an exclamation symbol always appears in the messages talking about the WhatsApp version, that doesn't influence it, right? And thank you in advance for your help and answers, it’s helping a lot! Thanks

**Reply by a5r0n** [reply]
> Encrypting a message for a large group is very slow. So we need to avoid it being encrypted by whatsmeow .
> The current solution is limiting the participants in this list 
> 
> https://github.com/tulir/whatsmeow/blob/350073db195c6b6fae832a5b420b5b67fd39f6be/send.go#L539-L542
> 
> just like this :
> ```go
> participants = []types.JID{ownID}
> ```

I'm not sure if that is a safe solution, you probably just make it easier for WhatsApp to detect and ban..

**Reply by ghost** [reply]
> > Encrypting a message for a large group is very slow. So we need to avoid it being encrypted by whatsmeow .
> > The current solution is limiting the participants in this list 
> > 
> > https://github.com/tulir/whatsmeow/blob/350073db195c6b6fae832a5b420b5b67fd39f6be/send.go#L539-L542
> > 
> > just like this :
> > ```go
> > participants = []types.JID{ownID}
> > ```
> 
> I'm not sure if that is a safe solution, you probably just make it easier for WhatsApp to detect and ban..
> 

Please let us know if you have a better solution.

**Reply by PBatista** [reply]
@JustSoursop  Regarding the issue of trying to forward messages to other devices, did you have anything in mind or was it just an idea to be implemented? I wanted to see if it would be possible to do this in the background or something like that.

**Comment by AlmogBaku** [comment]
How does it work on w/a web? Why its faster there?

**Comment by MartinSchere** [comment]
One major improvement we made was switching to a high-speed database for part of the store—specifically for group participants.

By adding Redis within the same VPC as the server and consolidating the retrieval into a single call for all participants, we saw a significant boost in performance.

**Reply by KuntilBogel** [reply]
HI, can you provide an example of what code u used, and/or changed to speed things up>

**Reply by MartinSchere** [reply]
I can't provide the exact example as it's company/proprietary code.

But you essentially want to save as many network roundtrips as possible, both to your cache as well as to WhatsApp. When sending to large groups, the core things you want to look out for are:
- User devices cache (the library will fetch all user devices for all participants in the group)
- Group cache (gets all participants)
- Session store (stores all keys to encrypt messages)

You want all those to be as fast as possible, ideally an MGET or MSET (if you use Redis) away.

**Reply by AlmogBaku** [reply]
That's a shame. A major part of the OSS magic is to contribute back to the community as it's something directly related to this repo and not to the core of your business.

**Reply by MartinSchere** [reply]
Here you go @AlmogBaku 
https://github.com/tulir/whatsmeow/pulls/954

**Comment by xAbdaziz** [comment, marked as answer]
Closing this discussion
I was able to achieve 1-2 sec sending speed in a group with 600+ members

After updating to PostgreSQL 18 with io_uring on and this commit https://github.com/tulir/whatsmeow/commit/c8bf4476d21c70058e1cd7219391d1edc2eae2e2
